nanomq/apps/webhook_test.c

```diff
@@ -0,0 +1,478 @@
+	}
+	if ((rv = nng_dial(req_sock, INPROC_URL, NULL, NNG_FLAG_NONBLOCK)) !=
+	    0) {
+		fatal("nng_dial(" INPROC_URL ")", rv);
+	}
+
+	// Get a suitable HTTP server instance.  This creates one
+	// if it doesn't already exist.
+	if ((rv = nng_http_server_hold(&server, url)) != 0) {
+		fatal("nng_http_server_hold", rv);
+	}
+
+	// Allocate the handler - we use a dynamic handler for REST
+	// using the function "rest_handle" declared above.
+	rv = nng_http_handler_alloc(&handler, url->u_path, rest_handle);
+	if (rv != 0) {
+		fatal("nng_http_handler_alloc", rv);
+	}
+
+	if ((rv = nng_http_handler_set_tree(handler)) != 0) {
+		fatal("nng_http_handler_set_tree", rv);
+	}
+
+	if ((rv = nng_http_handler_set_method(handler, NULL)) != 0) {
+		fatal("nng_http_handler_set_method", rv);
+	}
+	// We want to collect the body, and we (arbitrarily) limit this to
+	// 128KB.  The default limit is 1MB.  You can explicitly collect
+	// the data yourself with another HTTP read transaction by disabling
+	// this, but that's a lot of work, especially if you want to handle
+	// chunked transfers.
+	if ((rv = nng_http_handler_collect_body(handler, true, 1024 * 128)) !=
+	    0) {
+		fatal("nng_http_handler_collect_body", rv);
+	}
+	if ((rv = nng_http_server_add_handler(server, handler)) != 0) {
+		fatal("nng_http_handler_add_handler", rv);
+	}
+	if ((rv = nng_http_server_start(server)) != 0) {
+		printf("things go wrong\n");
+		fatal("nng_http_server_start", rv);
+	}
+
+	nng_url_free(url);
+}
+
+//
+// inproc_server - this just is a simple REP server that listens for
+// messages, and performs ROT13 on them before sending them.  This
+// doesn't have to be in the same process -- it is hear for demonstration
+// simplicity only.  (Most likely this would be somewhere else.)  Note
+// especially that this uses inproc, so nothing can get to it directly
+// from outside the process.
+//
+static void
+inproc_server(void *arg)
+{
+	nng_socket s;
+	int        rv;
+	nng_msg *  msg;
+
+	if (((rv = nng_rep0_open(&s)) != 0) ||
+	    ((rv = nng_listen(s, INPROC_URL, NULL, 0)) != 0)) {
+		fatal("unable to set up inproc", rv);
+	}
+	// This is simple enough that we don't need concurrency.  Plus it
+	// makes for an easier demo.
+	for (;;) {
+		char *body;
+		if ((rv = nng_recvmsg(s, &msg, 0)) != 0) {
+			fatal("inproc recvmsg", rv);
+		}
+		body = nng_msg_body(msg);
+		// printf("\tReceived: %s\n", (char *) body);
+		nng_msg_free(msg);
+		webhook_msg_cnt++;
+
+		char *res = "OK";
+		if ((rv = nng_send(s, res, strlen(res), 0)) != 0) {
+			fatal("inproc sendmsg", rv);
+		}
+	}
+}
+
+static conf*
+get_webhook_conf()
+{
+	conf               *nanomq_conf;
+	conf_http_header   *header;
+	conf_web_hook_rule *webhook_rule;
+	if ((nanomq_conf = nng_zalloc(sizeof(conf))) == NULL) {
+		fprintf(stderr,
+		    "Cannot allocate storge for configuration, quit\n");
+		exit(EXIT_FAILURE);
+	}
+	conf_init(nanomq_conf);
+	nanomq_conf->url                    = "nmq-tcp://0.0.0.0:1881";
```

Comment:
Don't forget to replace the port number here with your macro.

Reply:
this is basically a substitute of the configure file. Port is not a separate configurable items.

Reply:
So. If port 1881 is used. Are there any codes to resolve this problem?

Reply:
no, have to change the code anyway.